### PR TITLE
feat(router): Add payment link support

### PR DESCRIPTION
**Pull request by sahkal**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

### Additional Changes

- [ ] This PR modifies the API contract
- [x] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Screenshot 2023-09-07 at 10 35 24 PM
Screenshot 2023-09-07 at 10 35 44 PM
Screenshot 2023-09-07 at 10 51 30 PM

Screenshot 2023-10-09 at 9 57 39 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/payments.rs [hunk 1]
```diff
@@ -102,6 +102,11 @@ where
         )
         .await?;
 
+    println!(
```
Comment (ashokkjag) [inline crates/router/src/core/payments.rs:105]: remove prints and use debug logs

File: crates/router/src/core/payments/helpers.rs [hunk 2]
```diff
@@ -2060,27 +2060,47 @@ pub fn generate_mandate(
...
 
 // A function to manually authenticate the client secret with intent fulfillment time
-pub fn authenticate_client_secret(
+pub async fn authenticate_client_secret(
     request_client_secret: Option<&String>,
     payment_intent: &PaymentIntent,
     merchant_intent_fulfillment_time: Option<i64>,
+    db: &dyn StorageInterface,
 ) -> Result<(), errors::ApiErrorResponse> {
+    let payment_link_db = if let Some(payment_link_id) = &payment_intent.payment_link_id {
```
Comment (ashokkjag) [inline crates/router/src/core/payments/helpers.rs:2069]: Why not move this inside the client_secret verification check? It may take a db hit even with a wrong client secret?

File: crates/router/src/services/api.rs [hunk 3]
```diff
@@ -1265,3 +1281,30 @@ mod tests {
         assert_eq!(mime::APPLICATION_JSON.essence_str(), "application/json");
     }
 }
+
+pub fn build_payment_link_html(payment_link_data: PaymentLinkFormData) -> maud::Markup {
+    use maud::PreEscaped;
+    let hyper_loader_sdk = if payment_link_data.base_url == "http://localhost:8080" {
+        "http://localhost:9090/dist/HyperLoader.js".to_string()
+    } else {
+        "https://beta.hyperswitch.io/v1/HyperLoader.js".to_string()
```
Comment (ashokkjag) [inline crates/router/src/services/api.rs:1290]: move this to a config file

File: crates/router/src/core/payment_link.rs [hunk 4]
```diff
@@ -0,0 +1,834 @@
...
+    amount: String,
+    currency: String,
+    pub_key: String,
+    secret: String,
+    payment_id: String,
+) -> String {
+    format!(
+    "
+    <script>
+        window.__PAYMENT_DETAILS_STR = JSON.stringify({{
```
Comment (ashokkjag) [inline crates/router/src/core/payment_link.rs:96]: Move this outside to a file which you can read on startup? Otherwise minor css changes too will require a full re compile.

**Review by ashokkjag — COMMENTED** [review]
LGTM otherwise

File: crates/router/src/core/payments/helpers.rs [hunk 5]
```diff
@@ -2060,27 +2060,50 @@ pub fn generate_mandate(
 }
 
 // A function to manually authenticate the client secret with intent fulfillment time
-pub fn authenticate_client_secret(
+pub async fn authenticate_client_secret(
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/helpers.rs:2063]: do not make this function as async, let it just do the authentication and not any db calls. All the database calls can be made prior and data can be passed along.

File: crates/router/src/core/payments/helpers.rs [hunk 6]
```diff
@@ -2060,27 +2060,50 @@ pub fn generate_mandate(
...
+                            >= payment_link_db
+                                .fullfilment_time
+                                .unwrap_or(current_timestamp),
+                        || Err(errors::ApiErrorResponse::ClientSecretExpired),
+                    )
+                } else {
+                    let payment_intent_fulfillment_deadline =
+                        payment_intent.created_at.saturating_add(Duration::seconds(
+                            merchant_intent_fulfillment_time
+                                .unwrap_or(consts::DEFAULT_FULFILLMENT_TIME),
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/helpers.rs:2099]: The logic is dependent on the fulfillment time, which can either come from `payment_intent` or `payment_link` (which overrides the payment_intent fulfillment time), can you refactor this based on this idea?

File: crates/router/src/core/payments/operations/payment_create.rs [hunk 7]
```diff
@@ -65,6 +66,96 @@ impl<F: Send + Clone> GetTracker<F, PaymentData<F>, api::PaymentsRequest> for Pa
...
+        let (payment_link, payment_link_id) = if let Some(payment_link_object) =
+            &request.payment_link_object
+        {
+            if !request.confirm.unwrap_or(false) {
+                let created_at @ last_modified_at = Some(common_utils::date_time::now());
+                if created_at > payment_link_object.link_expiry {
+                    return Err(report!(errors::ApiErrorResponse::InvalidRequestData {
+                        message: "link_expiry time cannot be less than current time".to_string(),
+                    }));
+                }
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/operations/payment_create.rs:78]: If this validation is not dependent on any of the database object, add this to `validate_request` function

File: crates/router/src/core/payments/operations/payment_create.rs [hunk 8]
```diff
@@ -65,6 +66,96 @@ impl<F: Send + Clone> GetTracker<F, PaymentData<F>, api::PaymentsRequest> for Pa
...
+                (
+                    Some(payment_link_db.link_to_pay),
+                    Some(payment_link_db.payment_link_id),
+                )
+            } else {
+                (None, None)
+            }
+        } else {
+            (None, None)
+        };
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/operations/payment_create.rs:113]: Move this to a function

File: crates/router/src/core/payments/operations/payment_create.rs [hunk 9]
```diff
@@ -388,6 +481,11 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, api::PaymentsRequest> for Paymen
             .clone();
         let authorized_amount = payment_data.payment_attempt.amount;
 
+        println!(
+            "non updated payment intent sahkal {:?}",
+            payment_data.payment_intent.clone()
+        );
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/operations/payment_create.rs:487]: Remove the debug lines

File: crates/router/src/core/payments/operations/payment_link.rs
Comment (Narayanbhat166) [inline crates/router/src/core/payments/operations/payment_link.rs:1]: Is this file required?

**Review by Narayanbhat166 — CHANGES_REQUESTED** [review]

File: crates/api_models/src/payments.rs [hunk 10]
```diff
@@ -2880,3 +2883,44 @@ mod tests {
         )
     }
 }
+
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize, PartialEq, ToSchema)]
+pub struct PaymentLinkObject {
+    pub create_payment_link: bool,
```
[the remark this answers is not in the document]
Comment (sahkal) [reply crates/api_models/src/payments.rs:2889]: created this field as a dummy value while starting, removing it!

File: crates/api_models/src/payments.rs [hunk 11]
```diff
@@ -2880,3 +2883,44 @@ mod tests {
...
+#[derive(Default, Debug, serde::Deserialize, serde::Serialize, Clone, ToSchema)]
+pub struct RetrievePaymentLinkRequest {
+    pub payment_link_id: String,
+    pub client_secret: Option<String>,
+}
+
+#[derive(Clone, Debug, serde::Serialize, PartialEq, serde::Deserialize)]
+pub struct PaymentLinkResponse {
+    pub payment_link: Option<String>,
+    pub payment_link_id: Option<String>,
```
[the remark this answers is not in the document]
Comment (sahkal) [reply crates/api_models/src/payments.rs:2903]: yes, changing it to String required if PaymentLinkResponse is present!

File: crates/router/src/compatibility/wrap.rs [hunk 12]
```diff
@@ -130,6 +130,20 @@ where
...
+            match api::build_payment_link_html(*payment_link_data) {
+                Ok(rendered_html) => api::http_response_html_data(rendered_html),
+                Err(_) => api::http_response_err(
+                    r#"{
+                        "error": {
+                            "message": "Error while rendering payment link html page"
+                        }
+                    }"#,
+                ),
+            }
```
[the remark this answers is not in the document]
Comment (sahkal) [reply crates/router/src/compatibility/wrap.rs:144]: no, since we are using a tera macro. where rendering might fail while generating the template we cannot have a string response as we have incase of build_redirection_form.

File: crates/router/src/core/payments/operations/payment_link.rs
Comment (sahkal) [reply crates/router/src/core/payments/operations/payment_link.rs:1]: not now have deleted it!

File: config/development.toml [hunk 13]
```diff
@@ -428,3 +428,6 @@ apple_pay_ppc = "APPLE_PAY_PAYMENT_PROCESSING_CERTIFICATE"
 apple_pay_ppc_key = "APPLE_PAY_PAYMENT_PROCESSING_CERTIFICATE_KEY"
 apple_pay_merchant_cert = "APPLE_PAY_MERCHNAT_CERTIFICATE"
 apple_pay_merchant_cert_key = "APPLE_PAY_MERCHNAT_CERTIFICATE_KEY"
+
+[payment_link_sdk_url]
+sdk_url = "http://localhost:9090/dist/HyperLoader.js"
```
[the remark this answers is not in the document]
Comment (sahkal) [reply config/development.toml:433]: i used orca-elements repo for sdk setup!

File: crates/diesel_models/src/payment_link.rs [hunk 14]
```diff
@@ -0,0 +1,50 @@
+use diesel::{Identifiable, Insertable, Queryable};
+use serde::{self, Deserialize, Serialize};
+use time::PrimitiveDateTime;
+
+use crate::{enums as storage_enums, schema::payment_link};
+
+#[derive(Clone, Debug, Eq, PartialEq, Identifiable, Queryable, Serialize, Deserialize)]
+#[diesel(table_name = payment_link)]
+#[diesel(primary_key(payment_id))]
```
[the remark this answers is not in the document]
Comment (sahkal) [reply crates/diesel_models/src/payment_link.rs:9]: correct! will change it primary key to payment_link_id

**Review by Narayanbhat166 — CHANGES_REQUESTED** [review]

File: crates/router/src/routes/payment_link.rs [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+use super::app::AppState;
+use crate::{
+    core::payment_link::{self},
+    services::{api, authentication as auth},
+};
+
+pub async fn get_payment_link(
+    state: web::Data<AppState>,
+    req: actix_web::HttpRequest,
+    query_payload: web::Query<api_models::payments::RetrievePaymentLinkRequest>,
```
Comment (Narayanbhat166) [inline crates/router/src/routes/payment_link.rs:13]: add this to path param

**Review by Narayanbhat166 — DISMISSED** [review]

File: config/development.toml [hunk 16]
```diff
@@ -441,6 +441,9 @@ apple_pay_ppc_key = "APPLE_PAY_PAYMENT_PROCESSING_CERTIFICATE_KEY"
 apple_pay_merchant_cert = "APPLE_PAY_MERCHNAT_CERTIFICATE"
 apple_pay_merchant_cert_key = "APPLE_PAY_MERCHNAT_CERTIFICATE_KEY"
 
+[payment_link]
```
Comment (jarnura) [inline config/development.toml:444]: add the same config.example.toml with documentation.

File: crates/api_models/src/payments.rs [hunk 17]
```diff
@@ -2904,3 +2907,44 @@ mod tests {
...
 }
+
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize, PartialEq, ToSchema)]
+pub struct PaymentLinkObject {
+    #[serde(default, with = "common_utils::custom_serde::iso8601::option")]
+    pub link_expiry: Option<PrimitiveDateTime>,
+    pub merchant_custom_domain_name: Option<String>,
+}
+
+#[derive(Default, Debug, serde::Deserialize, serde::Serialize, Clone, ToSchema)]
```
Comment (jarnura) [inline crates/api_models/src/payments.rs:2918]: Remove serde::Serialize for Requests

File: crates/api_models/src/payments.rs [hunk 18]
```diff
@@ -2904,3 +2907,44 @@ mod tests {
...
+pub struct RetrievePaymentLinkRequest {
+    pub client_secret: Option<String>,
+}
+
+#[derive(Clone, Debug, serde::Serialize, PartialEq, serde::Deserialize, ToSchema)]
```
Comment (jarnura) [inline crates/api_models/src/payments.rs:2923]: Remove serde::Deserialize for Response

File: crates/api_models/src/payments.rs [hunk 19]
```diff
@@ -2904,3 +2907,44 @@ mod tests {
...
+pub struct PaymentLinkResponse {
+    pub link: String,
+    pub payment_link_id: String,
+}
+
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize, ToSchema)]
+pub struct RetrievePaymentLinkResponse {
```
Comment (jarnura) [inline crates/api_models/src/payments.rs:2930]: Remove serde::Deserialize for Response

File: crates/api_models/src/payments.rs [hunk 20]
```diff
@@ -2904,3 +2907,44 @@ mod tests {
...
+    #[serde(with = "common_utils::custom_serde::iso8601")]
+    pub created_at: PrimitiveDateTime,
+    #[serde(with = "common_utils::custom_serde::iso8601")]
+    pub last_modified_at: PrimitiveDateTime,
+    #[serde(default, with = "common_utils::custom_serde::iso8601::option")]
+    pub link_expiry: Option<PrimitiveDateTime>,
+}
+
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize, ToSchema)]
+pub struct PaymentLinkInitiateRequest {
```
Comment (jarnura) [inline crates/api_models/src/payments.rs:2947]: Remove serde::Serialize for Requests

**Review by jarnura — CHANGES_REQUESTED** [review]

**Review by Narayanbhat166 — APPROVED** [review]
